File: without-library/graph/dag/topological_sort.cpp
#include <iostream>
#include <vector>

using namespace std;

vector<int> adj[200005];
int in_degree[200005];

vector<int> kahn(int n) {
    for(int i=1; i<=n; i++) {
        for(size_t j=0; j<adj[i].size(); j++) {
            in_degree[adj[i][j]]++;
        }
    }
    
    vector<int> q;
    for(int i=1; i<=n; i++) {
        if(in_degree[i] == 0) q.push_back(i);
    }
    
    int head = 0;
    vector<int> topo;
    
    while(head < q.size()) {
        int u = q[head++];
        topo.push_back(u);
        
        for(size_t i=0; i<adj[u].size(); i++) {
            int v = adj[u][i];
            in_degree[v]--;
            if(in_degree[v] == 0) q.push_back(v);
        }
    }
    
    if(topo.size() < n) return {};
    return topo;
}

int main() {
    return 0;
}

File: without-library/graph/dag/topological_sort.py
import sys

class TopoSort:
    def __init__(self, n, adj):
        self.n = n
        self.adj = adj
        self.in_degree = [0] * (n + 1)
        
    def solve(self):
        for u in range(1, self.n + 1):
            for v in self.adj[u]:
                self.in_degree[v] += 1
                
        queue = []
        for i in range(1, self.n + 1):
            if self.in_degree[i] == 0:
                queue.append(i)
                
        head = 0
        topo = []
        
        while head < len(queue):
            u = queue[head]
            head += 1
            topo.append(u)
            
            for v in self.adj[u]:
                self.in_degree[v] -= 1
                if self.in_degree[v] == 0:
                    queue.append(v)
                    
        return topo if len(topo) == self.n else []

if __name__ == "__main__":
    pass
